cranelift-codegen/meta/src/gen_legalizer.rs

```diff
@@ -34,27 +35,61 @@ fn unwrap_inst(
     let iform = format_registry.get(inst.format);
 
     fmt.comment(format!(
-        "Unwrap {}",
+        "Unwrap fields from instruction format {}",
         def.to_comment_string(&transform.var_pool)
     ));
 
     // Extract the Var arguments.
     let arg_names = apply
         .args
         .iter()
-        .map(|arg| match arg.maybe_var() {
-            Some(var_index) => var_pool.get(var_index).name,
-            None => "_",
+        .enumerate()
+        .filter(|(arg_num, _)| {
+            // Variable args are specially handled after extracting args.
+            !inst.operands_in[*arg_num].is_varargs()
+        })
+        .map(|(arg_num, arg)| match &arg {
+            Expr::Var(var_index) => var_pool.get(*var_index).name,
+            Expr::Literal(_) => {
+                let n = inst.imm_opnums.iter().position(|&i| i == arg_num).unwrap();
+                iform.imm_fields[n].member
+            }
         })
         .collect::<Vec<_>>()
         .join(", ");
 
+    // May we need "args" in the values consumed by predicates?
+    let emit_args = iform.num_value_operands >= 1 || iform.has_value_list;
+
+    // We need a tuple:
+    // - if there's at least one value operand, then we emit a variable for the value, and the
+    // value list as args.
+    // - otherwise, if there's the count of immediate operands added to the presence of a value list exceeds one.
+    let need_tuple = iform.num_value_operands >= 1
+        || (inst
+            .operands_in
+            .iter()
+            .filter(|op| op.is_immediate())
+            .count()
+            + if iform.has_value_list { 1 } else { 0 })
+            > 1;
```

Comment:
Personally, I try to avoid lines that get broken up like this, because I find them difficult to comfortably parse at a glance.

This is more manual, but I think it's probably easier to read and still avoids the calculation in the short-circuit case:

```rust
let need_tuple = if iform.num_value_operands >= 1 {
    true
} else {
    let mut immediates = inst.operands_in.iter().filter(|op| op.is_immediate()).count();
    if iform.has_value_list {
        immediates += 1;
    }
    immediates > 1
};
```

Something like that.